deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.kt

```diff
@@ -0,0 +1,140 @@
+ *
+ *
+ * The output location is specified in the project's gradle file through
+ * [ProcessingEnvironment]'s compiler argument options by deepLinkDoc.output.
+ */
+internal class Documentor(private val processingEnv: XProcessingEnv) {
+    private val messager: XMessager = processingEnv.messager
+
+    @get:VisibleForTesting
+    var file: File? = initFile()
+
+    fun write(elements: List<DeepLinkAnnotatedElement>) {
+        val file = file ?: run {
+            messager.printMessage(
+                Diagnostic.Kind.NOTE,
+                "Output file is null, DeepLink doc not generated."
+            )
+            return
+        }
+        if (elements.isNullOrEmpty()) {
+            messager.printMessage(
+                Diagnostic.Kind.NOTE,
+                "No deep link, DeepLink doc not generated."
+            )
+            return
+        }
+        try {
+            PrintWriter(FileWriter(file), true).use { writer ->
+                val fileName = file.name
+                val extIndex = fileName.lastIndexOf(".")
+
+                // markdown writer if .md file extension is used
+                val docWriter: DocumetationWriter =
+                    if (extIndex >= 0 && extIndex < fileName.length &&
+                        fileName.substring(extIndex + 1).equals("md", ignoreCase = true)
+                    ) {
+                        MarkdownWriter()
+                    } else {
+                        // else default writer
+                        GenericWriter()
+                    }
+                docWriter.write(processingEnv, writer, elements)
+            }
+        } catch (e: IOException) {
+            messager.printMessage(
+                Diagnostic.Kind.ERROR,
+                "DeepLink doc not generated: " + e.message
+            )
+        }
+        messager.printMessage(Diagnostic.Kind.NOTE, "DeepLink doc generated at: " + file.path)
+    }
+
+    private fun initFile(): File? {
+        val path = processingEnv.options[DOC_OUTPUT_PROPERTY_NAME]
+        if (path == null || path.trim { it <= ' ' }.isEmpty()) {
+            messager.printMessage(
+                Diagnostic.Kind.NOTE,
+                "Output path not specified, DeepLink doc is not going to be generated."
+            )
+            return null
+        }
+        val file = File(path)
+        if (file.isDirectory) {
+            messager.printMessage(
+                Diagnostic.Kind.NOTE,
+                "Specify a file path at $DOC_OUTPUT_PROPERTY_NAME to generate deep link doc.",
+            )
+            return file
+        }
+        val parentDir = file.parentFile
+        if (!parentDir.exists() && !parentDir.mkdirs()) {
+            messager.printMessage(
+                Diagnostic.Kind.NOTE,
+                "Cannot create file specified at ${file.canonicalPath}."
+            )
+        }
+        return file
+    }
+
+    /**
+     * Implement this interface if you want to provide own documentation writer.
```

Comment:
Was this a feature of DLD that uses could implement their own document writer? if so do we need to remove `internal`?